Blog post/ migrate from ember-css-modules (#2687)
mainmatter/mainmatter.com: src/posts/2025-03-28-migrate-from-ember-css-modules.md

## About ember-scoped-css

Just like ember-css-modules, ember-scoped-css is an Ember addon that brings CSS modules to your Ember apps. It is a v2 addon that can be associated with a v1 addon - ember-scoped-css-compat - in charge of the compatibility with classic apps.

### I want to start using CSS modules in my Ember Vite app, should I use ember-scoped-css?

No, not necessarily; ember-scoped-css is not the only option. You could also follow [Vite documentation](https://vite.dev/guide/features#css-modules) directly or look into what the Ember community came up with. The advantage of ember-scoped-css is that it's quite similar to ember-css-modules in the way the implementation is structured, so it's a migration path we recommend to unblock your upgrade to Vite without changing drastically all your CSS.

> ### I use ember-css-modules and want to migrate to a Vite-compatible solution, why should I favor ember-scoped-css?

Author: I am not sure that's a good idea. I mean I am not against turning the question into an assertion, but the suggestion in your comment doesn't say clearly enough what we must say:
- Don't use ember-scoped-css if you don't use ember-css-modules in the first place
- Favor ember-scoped-css over any other solution if you do

That's what I need to hear as a developer reading this blog post.

> ### I use ember-css-modules and want to migrate to a Vite-compatible solution, why should I favor ember-scoped-css?

Reviewer: I think I agree with kevin on this but not with his suggestion 😂 I would make the sub header here just `The case for ember-scoped-css` and then flow the content from both paragraphs together under that one header. If this doesn't make sense we can discuss it on a call or something 👍

> ### I use ember-css-modules and want to migrate to a Vite-compatible solution, why should I favor ember-scoped-css?

Reviewer: A single header `Is ember-scoped-css the best choice?` with those two paragraphs under it would be perfect 👌

### CSS isolation by class rename

ember-css-modules and ember-scoped-css both allow you to isolate your components styles by creating a CSS file alongside your components templates. They are different, but they both work following the same general idea: "CSS isolation by class rename". To understand this approach, you can use two resources:

- The document [CSS Isolation](https://github.com/soxhub/ember-scoped-css/blob/main/docs/css-isolation.md) from ember-scoped-css repository.
- The article [Cookbook: migrate an existing Ember app to CSS modules](https://mainmatter.com/blog/2022/08/24/cookbook-ember-app-to-css-modules/) is a walkthrough to install ember-css-modules in an Ember application that used initially one global CSS file. It presents the introduction of CSS modules with a more "practical" angle that allow you to see clearly the "without / with" CSS modules.

> To sum it up, ember-css-modules and ember-scoped-css both provide you ways to have your `custom-class` renamed to `custom-class+sha` following an interpolation pattern. This way, this class becomes unique and applies only to this specific component.

Reviewer 2: ```suggestion
To sum it up, ember-css-modules and ember-scoped-css both provide ways to have your `custom-class` renamed to `custom-class+sha` following an interpolation pattern. This way, this class becomes unique and applies only to this specific component.
```

ember-css-modules renames the classes to `._custom-class_sha`, and ember-scoped-css renames the classes to `.custom-class_sha`. It's a minor difference that turns to be important in the migration process (because both renaming methods conflict). There are also a few major differences that will force you to rework your code while migrating from one to the other.

## Introducing the resource repository

> To help you with the migration, we have created [a GitHub repository](https://github.com/BlueCutOfficial/css-modules-to-scoped-css) that serves as resource material. The demo consists of a `6.2` app inspired by [ember-welcome-page](https://github.com/ember-cli/ember-welcome-page). A `<WelcomePageCopy />` component has been implemented directly in the application and has been reworked to illustrate ember-css-modules and ember-scoped-css features.

Reviewer 2: Can you move the repo to the Mainmatter org?

Author: Probably, I can check how to do this. I don't think it should block the blog post though, we can update the links later.

## From ember-css-modules to ember-scoped-css

✨ [**Click here to see the diff**](https://github.com/BlueCutOfficial/css-modules-to-scoped-css/compare/demo-ember-css-modules..demo-ember-scoped-css) ✨

### 1. `local-class` _versus_ `class`

ember-css-modules introduces the attribute `local-class` has a marker that identifies scoped classes. In a template, it's your responsability as a developer to assign to `local-class` the scoped classes that you defined in the corresponding CSS module, and to assign to `class` the global classes.

In ember-scoped-css, you only use the attribute `class`. If the class name is defined in the CSS module, it will be transformed, else it just stays what it is and the global styles apply.

### 2. Tag selectors: global _versus_ local

ember-css-modules is designed to rename class selectors. It doesn't rename tag selectors like `div`, `p`, `img`, `a`... even when they are defined in the CSS module of a component. These selectors are bundled as is in the final style and therefore the style applies globally. In that way, using tag selectors in components when using ember-css-modules requires caution.

> ember-scoped-css is a bit more intuitive on that field. The tag selectors are no longer global but scoped. When these selectors are used in the component's CSS, a class named after the `sha` is added to the corresponding DOM elements, and the syntax `selector.sha` is used CSS-side to scope the style only to the component. This way, tag selectors behave just like any class selector, which can bring clearness to the CSS module.

Reviewer 2: ```suggestion
ember-scoped-css is a bit more intuitive on that field. The tag selectors are no longer global but scoped. When these selectors are used in the component's CSS, a class named after the `sha` is added to the corresponding DOM elements, and the syntax `selector.sha` is used CSS-side to scope the style only to the component. This way, tag selectors behave just like any class selector, which can make the CSS module clearer.
```

```html
<!-- Final DOM -->
<ul class="e9accf110">
  <li class="e9accf110"></li>
</ul>
```

```css
/* Built CSS */
ul.e9accf110 > li.e9accf110 {
  padding-bottom: 0.5em;
  font-size: 1.1em;
}
```

### 3. Id selectors: weird _versus_ global

If you use ember-css-modules, you may know already that its behavior with ids is a bit buggy. It doesn't rename elements ids in the DOM, but it transforms them in the CSS. As result, you can't use HTML ids to style the component, the styles wouldn't be applied:

```html
<!-- Final DOM -->
<main id="ember-welcome-page-id-selector"></main>
```

```css
/* Built CSS */
#_ember-welcome-page-id-selector_f42zxn {
  /* Style for _ember-welcome-page-id-selector_f42zxn which doesn't exist in the DOM */
}
```

> ember-scoped-css, on the other hand, tries to be a bit smarter on that field. It considers the following HTML principle: the HTML id should be unique in the document. Therefore, ember-scoped-css doesn't transform ids at all and the style remains global.

Reviewer 2: ```suggestion
ember-scoped-css, on the other hand, tries to be a bit smarter on that front. It considers the following HTML principle: the HTML id should be unique in the document. Therefore, ember-scoped-css doesn't transform ids at all and the style remains global.
```

on that front, in that regard, on that aspect…

```diff
+ <style>
+   @layer utilities, components;
+ </style>

  <link integrity="" rel="stylesheet" href="{{rootURL}}assets/vendor.css">
  <link integrity="" rel="stylesheet" href="{{rootURL}}assets/css-modules-to-scoped-css.css">
```

## From Classic to Vite once ember-scoped-css is used

✨ [**Click here to see the diff**](https://github.com/BlueCutOfficial/css-modules-to-scoped-css/compare/demo-ember-scoped-css..demo-ember-scoped-css-vite) ✨

Once your classic app uses only ember-scoped-css without style regressions, your app is one step closer to Vite. If you don't have any other blocker to handle, then you can start building with Vite. [ember-vite-codemod](https://github.com/mainmatter/ember-vite-codemod) is here to help!

👉 After running the codemod, remove `ember-scoped-css-compat` from your dependencies. It was here only for the compatibility with the classic app.

👉 [Adapt ember-scoped-css configuration](https://github.com/soxhub/ember-scoped-css?tab=readme-ov-file#configuration). You can also have a look at [the corresponding PR](https://github.com/BlueCutOfficial/css-modules-to-scoped-css/pull/2) on this repository to see additional comments.

## Conclusion

> With the introduction of 3.28 support, even your oldest Ember apps are closer to Vite than they have ever been. Managing all the prerequisite to use Vite is also a very good way to pave a smoother upgrade path for the future of your app and adopt all the modern practices at your own pace. As long as the [Ember Initiative](/ember-initiative/) goes on, we will continue to work hard to keep Ember align with latest standards, smart and easy to use. Do you rely on v1 addons that don't belong to the top 100? Do you need guidance to make them compatible with Vite or find a migration path like the one presented in this blog post? If you'd like to help us help you, and improve Ember for the entire web, consider supporting Mainmatter's Ember Initiative : [get in touch with us](/contact/), spread the word, and follow our progress on this blog.

Reviewer 2: ```suggestion
With the introduction of 3.28 support, even your oldest Ember apps are closer to Vite than they have ever been. Managing all the prerequisite to use Vite is also a very good way to pave a smoother upgrade path for the future of your app and adopt all the modern practices at your own pace. As long as the [Ember Initiative](/ember-initiative/) goes on, we will continue to work hard to keep Ember in line with the latest standards, smart and easy to use.

Do you rely on v1 addons that don't belong to the top 100? Do you need guidance to make them compatible with Vite or find a migration path like the one presented in this blog post? If you'd like to help us help you, and improve Ember for the entire web, [support the Ember Initiative](/contact/), spread the word, and follow our progress on this blog.
```

Also:

> keep Ember in line with latest standards, smart and easy to use.
I'd say maybe "keep Ember easy to use and up-to-date with the latest web standards"? "smart" doesn't quite fit imo

Author: Just a small practical note @KevinBongart: don't hesitate to make _every_ suggestion an actual suggestion. You can reply to your own comment to leave several suggestions for the same sentence, and then I can select the one I want to add to the commit batch. It's a tiny but appreciable time-saving thing.

---
title: "Get ready for Vite, migrate from ember-css-modules"
authorHandle: academierenards
tags: [ember, embroider, vite]
bio: "Marine Dunstetter, Senior Software Engineer"
description: "Introducing the migration path from ember-css-modules to ember-scoped-scc"
autoOg: true
tagline: |
  <p>

>   Ember apps from 3.28 to latest can build with Vite. This is a great achievement for the whole Ember community which Mainmatter has led as part of the <a href="/ember-initiative/">Ember Initiative</a>. Popular community addons must get ready for the change - more than half already are. However, a few classic addons rely too much on classic-world semantic and should be replaced with a different Vite-compatible solution. ember-css-modules is one of them. Fortunately, the Ember Initiative team has worked out a migration path you can follow to pave your way to Vite.

Reviewer: ```suggestion
  Ember apps from version 3.28 to the latest can now be built with Vite. This is a great achievement for the whole Ember community and is a result of work from the Mainmatter <a href="/ember-initiative/">Ember Initiative</a>. Popular community addons must get ready for the change - more than half already are. However, a few classic addons rely too much on classic-build semantics and should be replaced with a different Vite-compatible solution. ember-css-modules is one of them. Fortunately, the Ember Initiative team has worked out a migration path you can follow to pave your way to Vite.
```

Reviewer: I wonder if we might want to tone down the selling in this initial paragraph a bit? 🙈 the paragraph will flow better if we remove the `This is a great achievement for the whole Ember community and is a result of work from the Mainmatter <a href="/ember-initiative/">Ember Initiative</a>` bit and it will still show that we are doing great work. What do we think?

Author: ```suggestion
  Ember apps from version 3.28 to the latest can now be built with Vite. Popular community addons must get ready for the change - more than half already are. However, a few classic addons rely too much on classic-build semantics and should be replaced with a different Vite-compatible solution. ember-css-modules is one of them. Fortunately, Mainmatter <a href="/ember-initiative/">Ember Initiative</a> team has worked out a migration path you can follow to pave your way to Vite.
```

  </p>
---

> Mainmatter started the [Ember Initiative](/ember-initiative/) and made Vite support our team's first goal. After one month of work, new Ember apps from `3.28` to latest can build with Vite. One blocker developers may encounter though relates to the addons they use. If your app depends on a classic addon that turns to be incompatible with Vite, you will be stuck in the classic-world. It's not the way of the Ember community to let this happen, it doesn't leave anyone behind, and we - the Ember Initiative team - are aligned with this philosophy. To prevent developers from being blocked by classic addons, we audited of the top 100 addons on [Ember Observer](https://emberobserver.com/) and started to sort them out: Which are already v2? Which classic ones should be compatible with Vite? Which ones need a bit of rework? And last but not least, which ones rely too much on classic-world and should be abandoned in favor of a different solution?

Reviewer: ```suggestion
Mainmatter started the [Ember Initiative](/ember-initiative/) and made Vite support our team's first goal. After one month of work, new Ember apps from version `3.28` to the latest can be built with Vite. One blocker developers may encounter though relates to the addons they use. If your app depends on a classic addon that is incompatible with Vite, you will be stuck in the classic-build world. This is not the way the Ember community usually operates, we don't like to leave anyone behind, and we - the Ember Initiative team - are aligned with this philosophy. To prevent developers from being blocked by classic addons, we audited all the top 100 addons on [Ember Observer](https://emberobserver.com/) and started to sort them out: Which are already v2 and fully compatible? Which classic ones should be compatible with Vite? Which ones need a bit of rework? And last but not least, which ones rely too much on classic-build semantics and should be abandoned in favor of a different solution?
```

[ember-css-modules](https://github.com/salsify/ember-css-modules) is a widely used addon that brings CSS modules to classic Ember apps. However, we identified it as an addon that should be replaced with a different solution before moving to `@embroider/vite`. The problem is that until a week ago, there was no out-of-the-box solution to migrate away from ember-css-modules. You would have to figure out a way all by yourself, but CSS management is so challenging, especially in bigger apps. Fear no more, we have paved the way for you: it goes through [ember-scoped-css](https://github.com/soxhub/ember-scoped-css). In this blog post, we will explain everything about the migration strategy we recommend and the type of refactoring you should expect.

## About ember-scoped-css

> Just like ember-css-modules, ember-scoped-css is an Ember addon that brings CSS modules to your Ember apps. It is a v2 addon that can be associated with a v1 addon - ember-scoped-css-compat - in charge of the compatibility with classic apps.

Reviewer: ```suggestion
Just like ember-css-modules, ember-scoped-css is an Ember addon that brings CSS modules to your Ember apps. It is a v2 addon that can be used in conjunction with a v1 addon - ember-scoped-css-compat - that brings the same functionality to classic apps.
```

### I want to start using CSS modules in my Ember Vite app, should I use ember-scoped-css?

> No, not necessarily; ember-scoped-css is not the only option. You could also follow [Vite documentation](https://vite.dev/guide/features#css-modules) directly or look into what the Ember community came up with. The advantage of ember-scoped-css is that it's quite similar to ember-css-modules in the way the implementation is structured, so it's a migration path we recommend to unblock your upgrade to Vite without changing drastically all your CSS.

Reviewer: ```suggestion
No, not necessarily; ember-scoped-css is not the only option. You could also follow [Vite documentation](https://vite.dev/guide/features#css-modules) directly or look into some other solutions the Ember community came up with. The advantage of ember-scoped-css is that it's quite similar to ember-css-modules in the way the implementation is structured, so it's a migration path we recommend to unblock your upgrade to Vite without drastically changing all your CSS.
```
